src/logger/Logger.js

```diff
@@ -0,0 +1,125 @@
+
+class Logger {
+    /* Function callback to invoke with a metric object */
+    onMetricLog: Function = console.log; // eslint-disable-line no-console
+
+    /* A list of unique events that have occurred */
+    uniqueEvents: Array<string> = [];
+
+    /**
+     * Composes a component and event name together to make a semi-unique tag.
+     *
+     * @param {string} component - The component name
+     * @param {string} name - The event namne
+     * @returns {string} A string containing the component and event name
+     */
+    createComponentEventName(component: string, name: string): string {
+        return `${component}::${name}`;
+    }
+
+    /**
+     * Checks to see if the specified event for the component has already been fired.
+     *
+     * @param {string} component - the component name
+     * @param {string} name - the event name
+     * @returns {boolean} True if the event has already been fired
+     */
+    hasLoggedEvent(component: string, name: string): boolean {
+        return this.uniqueEvents.includes(this.createComponentEventName(component, name));
+    }
+
+    /**
+     * Invokes the provided metric logging callback.
+     *
+     * @param {string} component - the name of the component
+     * @param {string} name  - the name of the event
+     * @param {Object} [data] - Additional data to apply to the event
+     */
+    logMetric(component: string, name: string, data?: Object = {}): void {
+        const event = {
+            component,
+            name,
+            timestamp: this.getTimestamp(),
+            sessionId: SESSION_ID,
+            type: 'metric',
+            ...data,
+        };
+
+        this.onMetricLog(event);
+    }
+
+    /**
+     * Logs a unique metric event. Prevents duplicate events from being logged in the session.
+     *
+     * @param {string} component - the name of the component
+     * @param {string} name  - the name of the event
+     * @param {Object} [data] - Additional data to apply to the event
+     */
+    logUniqueMetric(component: string, name: string, data?: Object): void {
+        if (this.hasLoggedEvent(component, name)) {
+            return;
+        }
+        this.logMetric(component, name, data);
+        this.uniqueEvents.push(this.createComponentEventName(component, name));
+    }
+
+    /**
+     * Log a timing related metric.
+     *
+     * @param {string} component - the name of the component
+     * @param {string} name - the event name
+     * @param {string} from - Timer mark for the start of the timer
+     * @param {string} to - Timer mark for the end of the timer
+     * @param {boolean} [isUnique] - If true, guarantees a unique event firing
+     * @returns {void}
+     */
+    logTimeMetric(component: string, name: string, from: string, to: string, isUnique?: boolean = false): void {
+        const tagInfo = {
+            from,
+            to,
+        };
+
+        if (isUnique) {
+            this.logUniqueMetric(component, name, tagInfo);
```

Comment:
Where are the high-resolution timestamps extracted from the `performance` API?

Reply:
The plan is to emit them via the `onMetric` prop, and have the parent app consume them.